packages/editor/src/lib/editor/Editor.ts

```diff
@@ -5197,6 +5210,33 @@ export class Editor extends EventEmitter<TLEventMap> {
 	deleteBinding(binding: TLBinding | TLBindingId) {
 		return this.deleteBindings([binding])
 	}
+	canBindShapes(
+		fromShape: TLShape | TLShapeId | undefined | null,
+		toShape: TLShape | TLShapeId,
+		type: string
+	): boolean
+	canBindShapes(
+		fromShape: TLShape | TLShapeId,
+		toShape: TLShape | TLShapeId | undefined | null,
+		type: string
+	): boolean
+	canBindShapes(
+		fromShape: TLShape | TLShapeId | undefined | null,
+		toShape: TLShape | TLShapeId | undefined | null,
+		type: string
+	): boolean {
+		const from = fromShape ? this.getShape(fromShape) : undefined
+		const to = toShape ? this.getShape(toShape) : undefined
+
+		const canBindFrom = from
+			? this.getShapeUtil(from).canBind({ shape: from, otherShape: to, direction: 'from', type })
+			: undefined
+		const canBindTo = to
+			? this.getShapeUtil(to).canBind({ shape: to, otherShape: from, direction: 'to', type })
+			: undefined
+
+		return canBindFrom !== false && canBindTo !== false
```

Comment:
this feels quite heavy. I wonder whether we need to actually refer to the instances of the shapes when deciding if it can bind,  for the time being can we just use the shape `type` and then maybe add support for deciding based on the actual shape instances later if anybody needs it?

So we could have e.g. `editor.canBindShapes({fromShapeType, toShapeType, bindingType})` and then the shape utils can implement the same interface `canBind({ fromShapeType, toShapeType, bindingType })` too.

Reply:
we need the actual instances in some cases (or at least... we need to go from ID to type first). i've moved those to the callsites though - i think i prefer this the way it is now.